core/src/main/java/com/linecorp/armeria/common/util/AsyncCloseableShutdownHooks.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
+
+import java.util.ArrayDeque;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A utility class for adding a task with closing a {@link AsyncCloseable} on shutdown.
+ */
+public final class AsyncCloseableShutdownHooks {
+
+    private static final Logger logger = LoggerFactory.getLogger(AsyncCloseableShutdownHooks.class);
+
+    private static final Map<AsyncCloseable, Queue<Runnable>> asyncCloseableOnShutdownTasks =
+            new LinkedHashMap<>();
+
+    private static boolean addedToShutdownHook;
+
+    /**
+     *　Adds a {@link Runnable} and a {@link AsyncCloseable} to the JVM shutdown hook.
+     */
+    public static void addClosingTask(@Nullable Runnable whenClosing,
+                                      AsyncCloseable asyncCloseable,
+                                      CompletableFuture<Void> closeFuture, String name) {
+        final Runnable task = () -> {
+            if (whenClosing != null) {
+                try {
+                    whenClosing.run();
+                } catch (Exception e) {
+                    logger.warn("whenClosing failed", e);
+                }
+            }
+            asyncCloseable.closeAsync().handle((unused, cause) -> {
+                if (cause != null) {
+                    logger.warn("Unexpected exception while closing a {}.", name, cause);
+                    closeFuture.completeExceptionally(cause);
+                } else {
+                    logger.debug("{} has been closed.", name);
+                    closeFuture.complete(null);
+                }
+                return null;
+            }).join();
```

Comment:
`.join()` will throw an exception, causing double notification, because you logged about the outcome in the above callback already. How about just doing `close()` and catching an exception to log? e.g.

```
try {
    asyncCloseable().close()
} catch (Throwable cause) { ... log here ... }
```